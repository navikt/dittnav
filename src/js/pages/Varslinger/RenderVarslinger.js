import React, { useEffect, useReducer } from 'react';
import { useLocation } from 'react-router-dom';
import useBeskjedStore from '../../hooks/useBeskjedStore';
import { varslingerReducer, initialVarslingerState } from '../../reducers/varslingerReducer';
import PageFrame from '../PageFrame';
import Varslinger from './Varslinger';
import DelayedSpinner from '../../components/DelayedSpinner';
import scroll from '../../utils/scroll';
import {
  ADD_BESKJEDER, ADD_OPPGAVER, ADD_INNBOKSER, ADD_INAKTIVE_BESKJEDER, ADD_INAKTIVE_OPPGAVER, ADD_INAKTIVE_INNBOKSER,
  ADD_INNLOGGING, SET_BESKJEDER_LOADING, SET_INAKTIVE_BESKJEDER_LOADING, BESKJEDER_ERROR, OPPGAVER_ERROR,
  INNBOKSER_ERROR, INAKTIVE_BESKJEDER_ERROR, INAKTIVE_OPPGAVER_ERROR, INNLOGGING_ERROR, INAKTIVE_INNBOKSER_ERROR,
} from '../../types/Actions';
import ApiType from '../../types/ApiType';
import useModal from '../../hooks/useModal';
import InnloggingsModal from '../../components/common/InnloggingsModal';

const RenderVarslinger = ({ api }) => {
  const [state, dispatch] = useReducer(varslingerReducer, initialVarslingerState);
  const [visModal, toggleModal, handleModal] = useModal();
  const store = useBeskjedStore();
  const location = useLocation();

  const dispatchResult = (type, result, _dispatch = dispatch) => (
    _dispatch({ type, payload: result })
  );

  const dispatchError = (type) => (
    dispatch({ type, payload: 'error.baksystemer' })
  );

  useEffect(
    () => {
      api.fetchBeskjeder()
        .then(([result, headers]) => {
          if (api.tokenExpiresSoon(headers)) {
            toggleModal();
          } else {
            dispatchResult(ADD_BESKJEDER, result, store.dispatch);
            dispatchResult(SET_BESKJEDER_LOADING, result);
          }
        })
        .catch(() => dispatchError(BESKJEDER_ERROR));

      api.fetchOppgaver()
        .then(([result, headers]) => {
          if (api.tokenExpiresSoon(headers)) {
            toggleModal();
          } else {
            dispatchResult(ADD_OPPGAVER, result);
          }
        })
        .catch(() => dispatchError(OPPGAVER_ERROR));

      api.fetchInnbokser()
        .then(([result, headers]) => {
          if (api.tokenExpiresSoon(headers)) {
            toggleModal();
          } else {
            dispatchResult(ADD_INNBOKSER, result);
          }
        })
        .catch(() => dispatchError(INNBOKSER_ERROR));

      api.fetchInaktiveBeskjeder()
        .then(([result, headers]) => {
          if (api.tokenExpiresSoon(headers)) {
            toggleModal();
          } else {
            dispatchResult(ADD_INAKTIVE_BESKJEDER, result, store.dispatch);
            dispatchResult(SET_INAKTIVE_BESKJEDER_LOADING, result);
          }
        })
        .catch(() => dispatchError(INAKTIVE_BESKJEDER_ERROR));

      api.fetchInaktiveOppgaver()
        .then(([result, headers]) => {
          if (api.tokenExpiresSoon(headers)) {
            toggleModal();
          } else {
            dispatchResult(ADD_INAKTIVE_OPPGAVER, result);
          }
        })
        .catch(() => dispatchError(INAKTIVE_OPPGAVER_ERROR));

      api.fetchInaktiveInnbokser()
        .then(([result, headers]) => {
          if (api.tokenExpiresSoon(headers)) {
            toggleModal();
          } else {
            dispatchResult(ADD_INAKTIVE_INNBOKSER, result);
          }
        })
        .catch(() => dispatchError(INAKTIVE_INNBOKSER_ERROR));

      api.fetchInnlogging()
        .then(([result]) => {
          if (!result.authenticated) {
            api.redirectToLogin();
          } else {
            dispatchResult(ADD_INNLOGGING, result);
          }
        })
        .catch(() => dispatchError(INNLOGGING_ERROR));
    }, [],
  );

  const areLoading = (key) => state[key].loading;
  const isLoading = Object.keys(state).some(areLoading);

  if (!isLoading && location.hash) {
    scroll(location.hash);
  }

  if (visModal) {
    return (<InnloggingsModal onClick={handleModal} isOpen />);
  }

  return (
    <>
      {isLoading
        ? <DelayedSpinner delay={500} spinnerClass="header-spinner spinner-container" />
        : (
          <PageFrame uniqueErrors={state.error}>
            <Varslinger
              oppgaver={state.oppgaver.data}
              innbokser={state.innbokser.data}
              inaktiveOppgaver={state.inaktiveOppgaver.data}
              inaktiveInnbokser={state.inaktiveInnbokser.data}
              innlogging={state.innlogging.data}
            />
          </PageFrame>
        )}
    </>
  );
};

RenderVarslinger.propTypes = {
  api: ApiType.isRequired,
};

export default RenderVarslinger;
