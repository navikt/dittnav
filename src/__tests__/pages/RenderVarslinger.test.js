import React from 'react';
import ReactDOM from 'react-dom';
import RenderVarslinger from 'js/pages/Varslinger/RenderVarslinger';
import wrapIntl from 'js/IntlTestHelper';

const ReactTestRenderer = require('react-test-renderer');

/* eslint-disable no-unused-vars */
const mockApi = () => (
  {
    fetchHendelser: () => new Promise((resolve, reject) => {}),
  }
);
/* eslint-enable no-unused-vars */

const flushPromises = () => (
  new Promise(resolve => setImmediate(resolve))
);

it('renders without crashing', () => {
  const div = document.createElement('div');

  ReactDOM.render(wrapIntl(<RenderVarslinger api={mockApi()} />), div);
  ReactDOM.unmountComponentAtNode(div);
});

it('expect Hendelser fetching', async () => {
  const api = mockApi();
  api.fetchHendelser = () => new Promise((resolve, reject) => { // eslint-disable-line no-unused-vars
    resolve(
      [
        {
          eventTidspunkt: '2019-11-27T17:51:26.001Z',
          eventId: '1874877086001',
          uid: '957de6ce-f44f-47de-84d2-639ab2684627',
          tekst: 'Vi mottok søknaden din 18. september 2019. Du kan følge med på statusen i Dine foreldrepenger.',
          link: 'https://enNyLenke',
          sistOppdatert: '2019-11-27T17:51:26.17575Z',
          type: 'BESKJED',
        },
        {
          eventTidspunkt: '2019-11-27T17:51:31.214Z',
          eventId: '1674877091214',
          tekst: 'Du har en sykemelding som må godkjennes',
          link: 'https://enNyLenke',
          sistOppdatert: '2019-11-27T17:51:31.414467Z',
          type: 'OPPGAVE',
        },
        {
          eventTidspunkt: '2019-11-27T17:51:31.214Z',
          eventId: '1574377091214',
          tekst: 'Svar fra veilederen din i innboksen: Hei, nå har jeg sjekket om...',
          link: 'https://enNyLenke',
          sistOppdatert: '2019-11-27T17:51:31.414467Z',
          type: 'INNBOKS',
        },
      ],
    );
  });
  const component = ReactTestRenderer.create(wrapIntl(<RenderVarslinger api={api} />));
  await flushPromises();

  expect(component.toJSON()).toMatchSnapshot();
});
