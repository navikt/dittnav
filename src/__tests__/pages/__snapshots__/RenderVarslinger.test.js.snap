// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect Brukernotifikasjoner fetching 1`] = `
<main
  role="main"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="maincontent side-innhold"
      >
        <div
          className="col-md-12"
          id="dittnav-main-container"
        >
          <div
            className="varslinger-tittel"
          >
            <h1
              className="typo-sidetittel"
            >
              <span>
                Beskjeder og oppgaver fra NAV
              </span>
            </h1>
          </div>
          <section
            className="infomeldinger-list"
          >
            <div
              className="alertbox"
            >
              <div
                className="alertstripe infomelding alertstripe--advarsel"
              >
                <span
                  className="alertstripe__ikon"
                >
                  <span
                    className="sr-only"
                  >
                    advarsel
                  </span>
                  <svg
                    focusable="false"
                    height="1.5em"
                    kind="advarsel-sirkel-fyll"
                    viewBox="0 0 24 24"
                    width="1.5em"
                  >
                    <g
                      fill="none"
                      fillRule="evenodd"
                    >
                      <path
                        d="M12.205-.004l-.214.002a12.225 12.225 0 0 0-8.517 3.659C1.179 5.977-.053 9.013.002 12.208c.115 6.613 5.296 11.793 11.795 11.793l.212-.002c6.726-.116 12.105-5.595 11.99-12.21C23.883 5.178 18.702-.003 12.204-.003z"
                        fill="#FFA733"
                        fillRule="nonzero"
                      />
                      <path
                        d="M12.027 19H12A1.499 1.499 0 0 1 11.973 16L12 16a1.501 1.501 0 0 1 .027 3z"
                        fill="#3E3832"
                      />
                      <path
                        d="M12 5a1 1 0 0 1 1 1v7a1 1 0 0 1-2 0V6a1 1 0 0 1 1-1z"
                        fill="#3E3832"
                        fillRule="nonzero"
                      />
                    </g>
                  </svg>
                </span>
                <div
                  className="typo-normal alertstripe__tekst"
                >
                  <h2
                    className="typo-undertittel"
                  >
                    <span>
                      Denne siden er under utvikling
                    </span>
                  </h2>
                  <p
                    className="typo-normal"
                  >
                    <span>
                      Her vil du kunne se alle tidligere meldinger fra NAV, men akkurat nå vises bare noen av beskjedene og oppgavene. Vi jobber med å forbedre siden og vil oppdatere den fortløpende. Du kan finne flere meldinger i 
                      <a
                        className="lenke"
                        href="http://localhost:9222/mininnboks"
                        id="alert-lenke-id"
                      >
                        innboksen
                      </a>
                       eller se informasjon om sakene dine i 
                      <a
                        className="lenke"
                        href="http://localhost:9222/saksoversikt"
                        id="alert-lenke-id"
                      >
                        Dine saker
                      </a>
                      .
                    </span>
                  </p>
                </div>
              </div>
            </div>
            <div
              className="aktive-varsler"
            />
            <div
              className="inaktive-varsler"
            />
            <div
              className="panel mininnboks-panel"
            >
              <p
                className="typo-normal"
              >
                <span>
                  Finner du ikke meldingen du leter etter? Den ligger kanskje i 
                  <a
                    className="lenke"
                    href="http://localhost:9222/mininnboks"
                    id="innboksmelding-id"
                  >
                    innboksen
                  </a>
                   din.
                </span>
              </p>
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</main>
`;
